libbeat/common/mapval/core.go

```diff
@@ -39,32 +40,41 @@ func Optional(id IsDef) IsDef {
 // Map is the type used to define schema definitions for Schema.
 type Map map[string]interface{}
 
+// Slice is a convenience []interface{} used to declare schema defs.
+type Slice []interface{}
+
 // Validator is the result of Schema and is run against the map you'd like to test.
-type Validator func(common.MapStr) *Results
+type Validator func(common.MapStr) (*Results, error)
```

Comment:
Good question, I think you found a bug. That error is supposed to occur if the validation can't complete. The errors on `Results` aren't processing errors, but rather validation result errors conveniently implementing `error`.

The bug I found is that an `IsDef` is allowed to return an error if it can't complete, but it wasn't properly being passed up to this return. I've patched that in 711d60a , and I've added tests for that case, and the other case of an invalid path being used, like "foo...bar", as opposed to "foo.bar".

Thinking through it, however, I suppose we can and should validate paths when creating the `Validator`, not after its created. OTOH, that means users have to handle errors at yet another stage. On the *other* hand, the `Schema` declaration could 'precompile' path strings to `Path` objects, which is more efficient for repeat execution (which we don't do currently).

WDYT?

Practically, it doesn't matter for our current uses.

Reply:
To clarify `Results.Errors()` returns invalid values according to the supplied schema, while the error you mentioned is more like an internal error where the validator couldn't finish execution.